### ISSUE TITLE
CPU: float32 is 3-5x faster than the bfloat16 default; and what should allowed_mass look like on a GPU?

### ISSUE THREAD

**Issue by MrJev** [OmniJev/PlayJev#1]
Hi — I reviewed PlayJev for a write-up on mrjev.com, at `2d7a028`. No GPU here, so I ran the harness on four CPU cores in the Playwright 1.63 image and then loaded `OmniJev/PlayJev-0.8B` on the same four cores. Two observations and one question; nothing here is a bug report.

**The harness held up to everything I threw at it.** Four headless pages, random policy: 67 env-steps/s on Snake, 131–177 on Floppy Bird. Episodes reproduced exactly — same seed plus the same action list gave byte-identical JPEG frames, scores and virtual-clock values across separate runs. (Two different seeds also gave me an identical short Snake episode, which `HARNESS.md` already explains: load-time randomness precedes the seed, so the opening position is fixed. Worth keeping that sentence where it is; it saved me filing a confused issue.)

I also checked the README table against `demo/results.json` rather than trying to reproduce it: all ten rows and all ten vs-teacher ratios match, 16 episodes each, `same_seed: true` throughout, mean 0.535 against the 0.53 quoted. Shipping the artifacts the table was computed from is not common, and the write-up running its own controls — random handover, inverted handover, the ablation that lost points on Mario and Racer, the Snake paragraph where agreement rose while the score fell — is the reason I'd point people at the results section before the code.

**A CPU note that might be worth a line in the README.** `requirements.txt` pins `flash-linear-attention`, `triton` and cu128 torch, so on CPU transformers falls back to the reference `chunk_gated_delta_rule` and `causal_conv1d_fn` kernels. That's expected. What surprised me is the dtype: the default `bfloat16` is emulated on CPU, and a single move took 59–110 s, while `PlayJevModel(..., device="cpu", dtype="float32")` did the same move in **20 s** — same answer to three decimals. If you ever expect people to poke at it without a GPU, "on CPU pass `dtype='float32'`" would save them a few minutes per frame. (Separately: I first fed it `games/flappy/screencap.png`, which is a 636×1222 page screenshot rather than a 448-long-side harness capture, and that pushed a move past five minutes and flattened the distribution — my mistake, but a reader could make it too, since that file sits in the games directory.)

**The question.** On every configuration I ran — bf16/plain, fp32/plain, bf16/chat, on real harness frames — `Decision.allowed_mass` came out around **1e-4**, while `top_token` was still ` B`:

```
frame 0060, fp32, plain:  choice=wait probs=[0.082, 0.918] conf=0.836 allowed_mass=0.000118 top_token=' B'
frame 0060, bf16, chat:   choice=wait probs=[0.051, 0.949] conf=0.897 allowed_mass=0.000102 top_token=' B'
```

The ratio between the letters looks right and is stable across dtype and template, so the readout works. But the docstring calls `allowed_mass` the "is the model answering at all?" diagnostic, and ~1e-4 implies a very flat full-vocabulary distribution for a checkpoint that was fully fine-tuned on this exact prompt. Is that what you see on the H200 as well, or should it be orders of magnitude higher there? If it should be high, then either the reference kernels change the picture more than transformers' "correct but much slower" wording suggests, or I'm holding something wrong — and either way a typical value in the model card would let people sanity-check their setup.

Thanks for publishing the harness, the teachers and the demo's result files alongside the weights. It's the most checkable thing I've reviewed this week.

**Comment by unikcc** [comment]
Hi, thank you for the attention and careful read. 

You are right about the docstring, we overlooked it and will fix it.

We checked allowed_mass on the GPU and it is just as small there, around 1e-5 on both bf16 and fp32, so your setup is fine. The training loss only shapes the letters relative to each other, and the decision is the softmax over the letter slots alone, so this share of the whole vocabulary does not affect it. We are considering pruning the vocabulary to a few hundred entries rather than the full set, close to Jev's option space, which may give a help for the performanc
 
We will also add the dtype="float32" note for CPU to the README.

Thanks for your feedback and suggestions.
